File: src/commands/gradle-intellij.ts
import {promises as fs} from 'fs'
import * as path from 'path'
import {GroovyGenerator} from '../files/groovy'
import {gradleHomeDir} from '../utils'

type Config = {
  token: string;
}

export class GradleIntellij {
  private readonly token: string

  constructor(config: Config) {
    this.token = config.token
  }

  generateInitScript(): string {
    const ge = new GroovyGenerator()
    ge.block("afterProject { proj ->", () => {
      ge.line("if (proj.plugins.findPlugin(%s) == null) return", "org.jetbrains.intellij")
      // configure publish repository
      ge.block("proj.tasks.publishPlugin {", () => {
        ge.line("token = %s", this.token)
      })
    })

    return ge.toString()
  }

  async configure(): Promise<void> {

    const init_d = path.join(gradleHomeDir(), "init.d")
    const initScriptPath = path.join(init_d, "gradle-intellij.gradle")
    await fs.mkdir(init_d, { recursive: true })
    // if file exists: throw error
    // eslint-disable-next-line github/no-then
    if (await fs.stat(initScriptPath).then(() => true, () => false)) {
      throw new Error("can't create gradle-intellij.gradle: exists")
    }
    await fs.writeFile(initScriptPath, this.generateInitScript(), { encoding: 'utf8' })
  }
  
  toString(): string {
    return 'gradle-intellij(with token)'
  }
}

File: src/commands/gradle_intellij.rs
use crate::utils::gradle::escape_groovy_string;
use crate::CmdResult;
use clap::Parser;

#[derive(Debug, Parser)]
#[command(no_binary_name = true)]
#[command(name = "prepare-gradle-intellij")]
/// Configure gradle globally to publish to intellij plugin portal
pub(crate) struct GradleIntellij {
    token: String,
}

impl GradleIntellij {
    fn generate_init_script(&self) -> String {
        format!(
            concat!(
                "afterProject {{ proj ->\n",
                "  if (proj.plugins.findPlugin(\"org.jetbrains.intellij\") == null) return\n",
                "  proj.tasks.publishPlugin {{\n",
                "    token = \"{token}\"\n",
                "  }}\n",
                "}}\n",
            ),
            token = escape_groovy_string(&self.token)
        )
    }

    pub async fn configure(&self) -> CmdResult {
        let mut path = crate::utils::gradle::gradle_home_dir();
        path.push("init.d/gradle-intellij.gradle");
        let path = path;

        crate::utils::write_to_new_file(path, self.generate_init_script().as_bytes())
            .await
            .expect("failed to create init script");

        ok!()
    }
}

#[test]
fn generated_init_script() {
    let intellij = GradleIntellij {
        token: "gradle-intellij-token-here".into(),
    };

    assert_eq!(
        intellij.generate_init_script(),
        concat!(
            "afterProject { proj ->\n",
            "  if (proj.plugins.findPlugin(\"org.jetbrains.intellij\") == null) return\n",
            "  proj.tasks.publishPlugin {\n",
            "    token = \"gradle-intellij-token-here\"\n",
            "  }\n",
            "}\n",
        )
    );
}
